File: src/dev/namtx/leetcode/DailyTemperatures.java
package dev.namtx.leetcode;

import java.util.Stack;

/**
 * https://leetcode.com/problems/daily-temperatures/
 * <p>
 * tags: monotonic stack
 */
public class DailyTemperatures {
    public static void main(String[] args) {
        System.out.println(new DailyTemperatures().dailyTemperatures(new int[]{73, 74, 75, 71, 69, 72, 76, 73}));
    }

    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> stack = new Stack<>();
        int[] ans = new int[temperatures.length];

        for (int i = temperatures.length - 1; i >= 0; i--) {
            while (!stack.isEmpty() && temperatures[i] >= temperatures[stack.peek()]) {
                stack.pop();
            }
            ans[i] = stack.isEmpty() ? 0 : stack.peek() - i;
            stack.push(i);
        }

        return ans;
    }
}

File: cpp/daily-temperatures.cpp
// https://leetcode.com/problems/daily-temperatures

#include <iostream>
#include <stack>
#include <vector>

using namespace std;

static auto _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  cout.tie(nullptr);
  return nullptr;
}();

class Solution {
public:
  vector<int> dailyTemperatures(vector<int> &temperatures) {
    stack<pair<int, int>> s;
    vector<int> ans(temperatures.size(), 0);

    for (int i = 0; i < temperatures.size(); i++) {
      int temp = temperatures[i];
      if (s.empty()) {
        s.push({i, temp});
      } else {
        if (temp > s.top().second) {
          while (!s.empty() && temp > s.top().second) {
            ans[s.top().first] = i - s.top().first;
            s.pop();
          }
        }
        s.push({i, temp});
      }
    }
    return ans;
  }
};
